react/src/Page.js

```diff
@@ -0,0 +1,96 @@
+import { useContext } from 'react';
+import { Navigate } from 'react-router-dom';
+import HeaderData from './HeaderData';
+import Auth from './utils/Auth';
+import Env from './utils/Env';
+import Client from './utils/Client';
+import Cookie from './utils/Cookie';
+
+// -------------------------------------------------------------------------------------------------
+// Page guards.
+// -------------------------------------------------------------------------------------------------
+
+// If the current environment (from the URL) is NOT a known environment according
+// to the envs.unique_annotated list in the auth record of the global header
+// data (from the React API /header endpoint), then redirect to the /env page.
+//
+function KnownEnvRequired({ children }) {
+    //
+    // Maybe just NoteLastUrl on AuthorizationRequired pages,
+    // i.e. all protected pages except the /login page.
+    // NoteLastUrl();
+    //
+    const [ header ] = useContext(HeaderData);
+    //
+    // TODO: More fully understand this next line added 2022-09-16.
+    // If not here then going to the /login page redirects to the /env page because /header API is still in progress.
+    // But thing that's a little confusing is we don't seem to need this kind of thing
+    // in AuthorizationRequired below.
+    //
+    if (header.loading) return children;
+    if (!Env.IsKnown(Env.Current(), header) ) {
+        return <Navigate to={Client.Path("/env")} replace />
+    }
+    else {
+        return children;
+    }
+}
+
+// If the user is NOT authenticated (i.e. logged in) OR is NOT authorized for the current
+// environment (i.e. if authenticated user is NOT allowed to access the environment from
+// the current URL, according to the allowed_envs list in the auth record of the global
+// header data (from the React API /header endpoint), then redirect to either the /login
+// page, if not authenticated, or to the /env page, if authenticated by not authorized.
+//
+function AuthorizationRequired({ children }) {
+    const [ header ] = useContext(HeaderData);
+    NoteLastUrl(header);
+    if (!Auth.IsLoggedIn(header)) {
+        return <Navigate to={Client.Path("/login")} replace />
+    }
+    else if (!Env.IsAllowed(Env.Current(), header)) {
+        return <Navigate to={Client.Path("/env")} replace />
+    }
+    else if (!Env.IsCurrentKnown(header)) {
```

Comment:
I guess I figure since this function is namespaced to `Env` then it should be obvious; all `Env` functions are environment (meaning the environment in the URL, e.g. cgap-supertest, or hotseat) related.

Reply:
Actually, I removed this, it didn't add much value, clearer just to use Env.IsKnown(Env.Current(), header).